packages/network-controller/src/rpc-service/rpc-service.ts

```diff
@@ -485,26 +488,55 @@ export class RpcService implements AbstractRpcService {
     } catch (error) {
       if (error instanceof HttpError) {
         const status = error.httpStatus;
-        if (status === 405) {
-          throw rpcErrors.methodNotFound();
+        if (status === 401) {
+          throw new JsonRpcError(-33100, 'Unauthorized.', {
+            httpStatus: status,
+          });
+        }
+        if (status === 402 || status === 404) {
+          throw rpcErrors.resourceUnavailable({
+            message: `RPC endpoint server error (HTTP ${status})`,
+            data: {
+              httpStatus: status,
+            },
+          });
+        }
+        if (status === 405 || status === 501) {
+          throw rpcErrors.methodNotFound({
+            data: {
+              httpStatus: status,
+            },
+          });
         }
         if (status === 429) {
           throw rpcErrors.limitExceeded({
             message: 'Request is being rate limited.',
+            data: {
+              httpStatus: status,
+            },
           });
         }
-        if (status === 503 || status === 504) {
-          throw rpcErrors.internal({
-            message:
-              'Gateway timeout. The request took too long to process. This can happen when querying logs over too wide a block range.',
+        if (status >= 500 && status < 600) {
+          const message =
+            status === 503 || status === 504
```

Comment:
The description for HTTP status 503 is ["Service Unavailable"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/503), so should we interpret it as `resourceUnavailable` instead of a timeout error?

Reply:
I think we should drop "Gateway timeout" from the message altogether honestly. Elliot I believe this is something you added in the process of migrating this error handling from the middleware, perhaps accidentally.

Reply:
Ah yes, it seems handling 503 as a timeout error, as well as the custom timeout error message, were originally in the middleware packages:

- https://github.com/MetaMask/eth-json-rpc-middleware/blob/ccaa4c91423437238d3c81a8201f0fa8a5ebadad/src/fetch.ts#L248
- https://github.com/MetaMask/eth-json-rpc-infura/blob/f10fd06512498c0eff8a377edd66b49d96b4b50b/src/create-infura-middleware.ts#L298

Reply:
The reason that I don't think "Gateway timeout" is appropriate here is that that's a common error message for 503 errors, but it's not the explanation for all such errors. We're assuming it's a gateway timeout, and that could be extremely misleading for devs if that's not really what's happening.

Reply:
So it sounds like we could simplify this to the following. Is this what we want to do?

``` typescript
if (status >= 500) {
  throw rpcErrors.resourceUnavailable({
    message: `RPC endpoint server error (HTTP ${status})`,
    data: {
      httpStatus: status,
    },
  });
}
```

Reply:
Yeah, personally I think that's sufficient, and less misleading.

Maybe would be nice to include more error details in the `data` property, but it's not clear how to do that simply and safely and without too much overhead. And I figure the dev could always open dev tools and see it directly if they really want to know more.